src/mmu/paging.c

```diff
@@ -65,116 +64,163 @@ page_table_t* next_table_address(page_table_t* table, uint64_t index)
-    uint64_t frame_number = pointed_frame(p3->entries[p3_index(page_number)]);
-    MMU_DEBUG("1GB huge page=%u frame=%u", page_number, frame_number);
+  if (p3 == NULL) {
+    MMU_DEBUG("did not find p3 (%p), returning NULL", p3);
+    return out_value;
+  }
 
-    if (frame_number % (PAGE_ENTRIES * PAGE_ENTRIES) == 0) {
-      frame_number +=
-        p2_index(page_number) * PAGE_ENTRIES + p1_index(page_number);
+  page_entry_t *p3_entry = &p3->entries[p3_index(page_number)];
 
-      return frame_number;
-    }
+  opt_uint64_t start_frame = pointed_frame(*p3_entry);
 
-    PANIC("misaligned 1GB page=%u", page_number);
-  }
+  if(start_frame.is_valid && p3_entry->huge_page == 1) {
+    MMU_DEBUG("1GB huge page=%u frame=%u", page_number, start_frame.value);
+
+    if(start_frame.value % (PAGE_ENTRIES * PAGE_ENTRIES) == 0) {
+      out_value.value = start_frame.value + p2_index(page_number) * PAGE_ENTRIES + p1_index(page_number);
+      out_value.is_valid = true;
 
-  if (p3 == 0) {
-    MMU_DEBUG("did not find p3 (%p), returning 0", p3);
-    return 0;
+      return out_value;
+    } else {
+      PANIC("misaligned 1GB page=%u", page_number);
+    }
   }
 
   page_table_t* p2 = next_table_address(p3, p3_index(page_number));
 
-  if (p3->entries[p3_index(page_number)].huge_page) {
-    uint64_t frame_number = pointed_frame(p2->entries[p2_index(page_number)]);
-    MMU_DEBUG("2MB huge page=%u frame=%u", page_number, frame_number);
+  if(p2 != NULL) {
 
-    if (frame_number % PAGE_ENTRIES == 0) {
-      frame_number += p1_index(page_number);
+    page_entry_t *p2_entry = &p2->entries[p2_index(page_number)];
 
-      return frame_number;
-    }
+    start_frame = pointed_frame(*p2_entry);
 
-    PANIC("misaligned 2MB page=%u", page_number);
-  }
+    if(start_frame.is_valid && p2_entry->huge_page == 1) {
+      MMU_DEBUG("2MB huge page=%u frame=%u", page_number, start_frame.value);
+
+      if(start_frame.value % PAGE_ENTRIES == 0) {
+        out_value.value = start_frame.value + p1_index(page_number);
+        out_value.is_valid = true;
 
-  if (p2 == 0) {
-    MMU_DEBUG("did not find p2 (%p), returning 0", p2);
-    return 0;
+        return out_value;
+      } else {
+        PANIC("misaligned 2MB page=%u", page_number);
+      }
+    }
+  } else {
+    MMU_DEBUG("did not find p2 (%p), returning NULL", p2);
+    return out_value;
   }
 
   page_table_t* p1 = next_table_address(p2, p2_index(page_number));
 
-  if (p1 == 0) {
-    MMU_DEBUG("did not find p1 (%p), returning 0", p1);
-    return 0;
+  if (p1 == NULL) {
+    MMU_DEBUG("did not find p1 (%p), returning NULL", p1);
+    return out_value;
   }
 
-  return pointed_frame(p1->entries[p1_index(page_number)]);
+  opt_uint64_t frame = pointed_frame(p1->entries[p1_index(page_number)]);
+
+  if(!frame.is_valid) {
+
+    PANIC("misaligned p1 page=%u", page_number);
+  }
+
+  return frame;
 }
 
 uint64_t p4_index(uint64_t page)
 {
-  return (page >> 27) & 0777;
+  return (page >> 27) & 0x1FF;
```

Comment:
Rust OS book states `0o777`, which is octal `777`. So this probably means your page indices were incorrect. Changed to hex.

Reply:
My apologies, I've just learned that `0777` is actually octal `777`. Please discregard this comment.